diff --git a/lib/facter/operatingsystem/ciscolinux.rb b/lib/facter/operatingsystem/ciscolinux.rb
new file mode 100644
index 0000000..41ebf87
--- /dev/null
+++ b/lib/facter/operatingsystem/ciscolinux.rb
@@ -0,0 +1,34 @@
+require 'facter/util/ciscoos'
+require 'facter/operatingsystem/linux'
+
+module Facter
+  module Operatingsystem
+    class CiscoLinux < Linux
+
+      def initialize(file)
+        @cisco_info = Facter::Util::CiscoOS.cisco_release(file)
+      end
+
+      def get_operatingsystem
+        @operatingsystem ||= @cisco_info['ID']
+        @operatingsystem
+      end
+
+      def get_osfamily
+        @osfamily ||= @cisco_info['ID_LIKE']
+        @osfamily
+      end
+
+      def get_operatingsystemrelease
+        @operatingsystemrelease ||= @cisco_info['VERSION']
+        @operatingsystemrelease
+      end
+
+      def get_operatingsystemmajorrelease
+        if operatingsystemrelease = get_operatingsystemrelease
+          operatingsystemrelease.split("(").first
+        end
+      end
+    end
+  end
+end
diff --git a/lib/facter/operatingsystem/implementation.rb b/lib/facter/operatingsystem/implementation.rb
index b453308..42950fd 100644
--- a/lib/facter/operatingsystem/implementation.rb
+++ b/lib/facter/operatingsystem/implementation.rb
@@ -1,6 +1,7 @@
 require 'facter/operatingsystem/base'
 require 'facter/operatingsystem/cumuluslinux'
 require 'facter/operatingsystem/linux'
+require 'facter/operatingsystem/ciscolinux'
 require 'facter/operatingsystem/sunos'
 require 'facter/operatingsystem/vmkernel'
 require 'facter/operatingsystem/windows'
@@ -13,6 +14,9 @@ module Facter
         release_info = Facter::Util::Operatingsystem.os_release
         if release_info['NAME'] == "Cumulus Linux"
           Facter::Operatingsystem::CumulusLinux.new
+        elsif !release_info['CISCO_RELEASE_INFO'].nil? &&
+              File.readable?(release_info['CISCO_RELEASE_INFO'])
+          Facter::Operatingsystem::CiscoLinux.new(release_info['CISCO_RELEASE_INFO'])
         else
           Facter::Operatingsystem::Linux.new
         end
diff --git a/lib/facter/util/ciscoos.rb b/lib/facter/util/ciscoos.rb
new file mode 100644
index 0000000..63ddbf1
--- /dev/null
+++ b/lib/facter/util/ciscoos.rb
@@ -0,0 +1,22 @@
+module Facter
+  module Util
+    module CiscoOS
+
+      # @file follows the same format as os-release
+      # @see http://www.freedesktop.org/software/systemd/man/os-release.html
+      def self.cisco_release(file)
+        values = {}
+
+        if File.readable?(file)
+          File.readlines(file).each do |line|
+            if (match = line.match(/^(\w+)=["']?(.+?)["']?$/))
+              values[match[1]] = match[2]
+            end
+          end
+        end
+
+        values
+      end
+    end
+  end
+end
